### PR TITLE
WIP: add spell & rpc tests

### PR DESCRIPTION
**Pull request by ilinzweilin**

### CODE REVIEW

File: src/spell.sol [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        // Init RWA-001 in Vat
+        VatAbstract(MCD_VAT).init(ilk);
+        // Init RWA-001 in Jug
+        JugAbstract(MCD_JUG).init(ilk);
+
+        // Allow RWA-001 Join to modify Vat registry
+        VatAbstract(MCD_VAT).rely(MCD_JOIN_NS2DRP_A);
+
+        // Allow RwaLiquidationOracle to modify Vat registry
+        VatAbstract(MCD_VAT).rely(MIP21_LIQUIDATION_ORACLE);
```
Comment (lucasvo) [inline src/spell.sol]: This is not needed AFAIK as it already is relied.

File: src/spell.sol [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        RwaLiquidationLike(MIP21_LIQUIDATION_ORACLE).init(
+            ilk, NS2DRP_A_INITIAL_PRICE, DOC, 300
+        );
+        (,address pip,,) = RwaLiquidationLike(MIP21_LIQUIDATION_ORACLE).ilks(ilk);
+        CHANGELOG.setAddress("PIP_NS2DRP", pip);
+
+        // Set price feed for NS2DRP
+        SpotAbstract(MCD_SPOT).file(ilk, "pip", pip);
+
+        // Init RWA-001 in Vat
```
Comment (lucasvo) [inline src/spell.sol]: replace all RWA-001 with NS2DRP

File: src/spell.sol [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    uint256 constant NS2DRP_A_INITIAL_DC    = 5 * MILLION * RAD; 
+    uint256 constant NS2DRP_A_INITIAL_PRICE = 1060 * WAD; // value ???
+
+    // MIP13c3-SP4 Declaration of Intent & Commercial Points -
+    // Off-Chain Asset Backed Lender to onboard Real World Assets
+    // as Collateral for a DAI loan
+    //
+    // https://ipfs.io/ipfs/QmdmAUTU3sd9VkdfTZNQM6krc9jsKgF2pz7W1qvvfJo1xk
+    string constant DOC = "QmdmAUTU3sd9VkdfTZNQM6krc9jsKgF2pz7W1qvvfJo1xk";
```
Comment (lucasvo) [inline src/spell.sol]: TODO: I'll provide a new hash later

File: src/spell.sol [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+    // precision
+    uint256 constant public THOUSAND = 10 ** 3;
+    uint256 constant public MILLION  = 10 ** 6;
+    uint256 constant public WAD      = 10 ** 18;
+    uint256 constant public RAY      = 10 ** 27;
+    uint256 constant public RAD      = 10 ** 45;
+
+    uint256 constant NS2DRP_A_INITIAL_DC    = 5 * MILLION * RAD; 
+    uint256 constant NS2DRP_A_INITIAL_PRICE = 1060 * WAD; // value ???
```
Comment (lucasvo) [inline src/spell.sol]: needs to be equal to DC+1yr of stability fees

File: src/spell.sol [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+    // MIP13c3-SP4 Declaration of Intent & Commercial Points -
+    // Off-Chain Asset Backed Lender to onboard Real World Assets
+    // as Collateral for a DAI loan
+    //
+    // https://ipfs.io/ipfs/QmdmAUTU3sd9VkdfTZNQM6krc9jsKgF2pz7W1qvvfJo1xk
```
Comment (lucasvo) [inline src/spell.sol]:
```suggestion
    // https://ipfs.io/ipfs/QmSwZzhzFgsbduBxR4hqCavDWPjvAHbNiqarj1fbTwpevR
```

File: src/spell.sol (hunk 3 above)
Comment (lucasvo) [reply src/spell.sol]:
```suggestion
    string constant DOC = "QmSwZzhzFgsbduBxR4hqCavDWPjvAHbNiqarj1fbTwpevR";
```

File: src/spell.sol [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    ChainlogAbstract constant CHANGELOG =
+        ChainlogAbstract(0xdA0Ab1e0017DEbCd72Be8599041a2aa3bA7e740F);
+
+    address constant NS2DRP_OPERATOR           = 0x27EfE12d1bede473960859E85375FaB75F4C9ffa;
+    address constant NS2DRP_GEM                = 0x46b25B14C572670F01De2453B03e001BAcAFF438;
+    address constant MCD_JOIN_NS2DRP_A         = 0x56Bc1E875cD1e557B90441ce7402CA2cB88D3819;
+    address constant NS2DRP_A_URN              = 0xB1d6C9079CD81e96e1719f1c22F5Fa4285f4e031;
+    address constant NS2DRP_A_INPUT_CONDUIT    = 0x27EfE12d1bede473960859E85375FaB75F4C9ffa;
+    address constant NS2DRP_A_OUTPUT_CONDUIT   = 0x27EfE12d1bede473960859E85375FaB75F4C9ffa;
+    address constant NS2DRP_LIQUIDATION_ORACLE  = 0xbB24f0c5e50EEA2f19CE996F4dcA6Ce216b58114;
```
Comment (godsflaw) [inline src/spell.sol]:
```suggestion
    address constant NS2DRP_LIQUIDATION_ORACLE  = 0x2881c5dF65A8D81e38f7636122aFb456514804CC;
```

File: src/spell.sol [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    // precision
+    uint256 constant public THOUSAND = 10 ** 3;
+    uint256 constant public MILLION  = 10 ** 6;
+    uint256 constant public WAD      = 10 ** 18;
+    uint256 constant public RAY      = 10 ** 27;
+    uint256 constant public RAD      = 10 ** 45;
+
+    uint256 constant NS2DRP_A_INITIAL_DC    = 5 * MILLION * RAD; 
+    // CreditLine + 2 years of stability fee
+    uint256 constant NS2DRP_A_INITIAL_PRICE = 5366480 * WAD; // 5,366,480
```
Comment (godsflaw) [inline src/spell.sol]:
```suggestion
    uint256 constant NS2DRP_A_INITIAL_PRICE = 5_366_480 * WAD; // 5,366,480
```

File: src/spell.sol [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+
+        // NS2DRP-A collateral deploy
+
+        // Set ilk bytes32 variable
+        bytes32 ilk = "NS2DRP-A";
+        
+        // add NS2DRP contract to the changelog
+        CHANGELOG.setAddress("NS2DRP", NS2DRP_GEM);
+        CHANGELOG.setAddress("MCD_JOIN_NS2DRP_A", MCD_JOIN_NS2DRP_A);
+        CHANGELOG.setAddress("NS2DRP_LIQUIDATION_ORACLE", NS2DRP_LIQUIDATION_ORACLE);
```
Comment (godsflaw) [inline src/spell.sol]:
```suggestion
```

File: src/spell.sol [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+
+        // Sanity checks
+        require(GemJoinAbstract(MCD_JOIN_NS2DRP_A).vat() == MCD_VAT, "join-vat-not-match");
+        require(GemJoinAbstract(MCD_JOIN_NS2DRP_A).ilk() == ilk, "join-ilk-not-match");
+        require(GemJoinAbstract(MCD_JOIN_NS2DRP_A).gem() == NS2DRP_GEM, "join-gem-not-match");
+        require(GemJoinAbstract(MCD_JOIN_NS2DRP_A).dec() == DSTokenAbstract(NS2DRP_GEM).decimals(), "join-dec-not-match");
+
+        // init the RwaLiquidationOracle
+        // doc: "doc"
```
Comment (godsflaw) [inline src/spell.sol]: make IPFS hash

File: src/spell.sol [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+        // tau: 5 minutes
+        RwaLiquidationLike(NS2DRP_LIQUIDATION_ORACLE).init(
```
Comment (godsflaw) [inline src/spell.sol]: change

File: src/spell.sol [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+            ilk, NS2DRP_A_INITIAL_PRICE, DOC, 300
+        );
+        (,address pip,,) = RwaLiquidationLike(NS2DRP_LIQUIDATION_ORACLE).ilks(ilk);
```
Comment (godsflaw) [inline src/spell.sol]: change

File: src/spell.sol [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+        // Init NS2DRP in Vat
+        VatAbstract(MCD_VAT).init(ilk);
+        // Init NS2DRP in Jug
+        JugAbstract(MCD_JUG).init(ilk);
+
+        // Allow NS2DRP Join to modify Vat registry
+        VatAbstract(MCD_VAT).rely(MCD_JOIN_NS2DRP_A);
+
+        // Allow RwaLiquidationOracle to modify Vat registry
+        VatAbstract(MCD_VAT).rely(NS2DRP_LIQUIDATION_ORACLE);
```
Comment (godsflaw) [inline src/spell.sol]: change

File: src/spell.sol [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+        VatAbstract(MCD_VAT).file("Line", VatAbstract(MCD_VAT).Line() + NS2DRP_A_INITIAL_DC);
+
+        // No dust
+        // VatAbstract(MCD_VAT).file(ilk, "dust", 0)
+
+        // 3.6% stability fee
+        JugAbstract(MCD_JUG).file(ilk, "duty", NS2DRP_THREEPOINTSIX_PERCENT_RATE);
+
+        // Set the NS2DRP-A min collateralization ratio (e.g. 105% => X = 105)
+        SpotAbstract(MCD_SPOT).file(ilk, "mat", 105 * RAY / 100);
```
Comment (godsflaw) [inline src/spell.sol]: change to 100%

File: src/test/addresses_kovan.sol [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        addr["MCD_JOIN_UNIV2DAIETH_A"]   = 0x03f18d97D25c13FecB15aBee143276D3bD2742De;
+        addr["MCD_FLIP_UNIV2DAIETH_A"]   = 0x0B6C3512C8D4300d566b286FC4a554dAC217AaA6;
+        addr["PROXY_PAUSE_ACTIONS"]      = 0x7c52826c1efEAE3199BDBe68e3916CC3eA222E29;
+        addr["PROXY_DEPLOYER"]           = 0xA9fCcB07DD3f774d5b9d02e99DE1a27f47F91189;
+        addr["NS2DRP"]                   = 0x46b25B14C572670F01De2453B03e001BAcAFF438;
+        addr["MCD_JOIN_NS2DRP_A"]        = 0x56Bc1E875cD1e557B90441ce7402CA2cB88D3819;
+        addr["NS2DRP_A_URN"]             = 0xB1d6C9079CD81e96e1719f1c22F5Fa4285f4e031;
+        addr["NS2DRP_A_INPUT_CONDUIT"]   = 0x27EfE12d1bede473960859E85375FaB75F4C9ffa;
+        addr["NS2DRP_A_OUTPUT_CONDUIT"]  = 0x27EfE12d1bede473960859E85375FaB75F4C9ffa;
+        addr["NS2DRP_LIQUIDATION_ORACLE"] = 0xbB24f0c5e50EEA2f19CE996F4dcA6Ce216b58114;
```
Comment (godsflaw) [inline src/test/addresses_kovan.sol]: change

**Review by godsflaw — APPROVED** [review]
LGTM

**Pull request merged**